src/Grpc.AspNetCore.Server/Internal/PipeExtensions.cs

```diff
@@ -0,0 +1,238 @@
+
+            pipeWriter.Advance(HeaderSize);
+        }
+
+        private static void EncodeMessageLength(int messageLength, Span<byte> destination)
+        {
+            Debug.Assert(destination.Length >= MessageDelimiterSize, "Buffer too small to encode message length.");
+
+            BinaryPrimitives.WriteUInt32BigEndian(destination, (uint)messageLength);
+        }
+
+        private static int DecodeMessageLength(ReadOnlySpan<byte> buffer)
+        {
+            Debug.Assert(buffer.Length >= MessageDelimiterSize, "Buffer too small to decode message length.");
+
+            var result = BinaryPrimitives.ReadUInt32BigEndian(buffer);
+
+            if (result > int.MaxValue)
+            {
+                throw new IOException("Message too large: " + result);
+            }
+
+            return (int)result;
+        }
+
+        private static bool TryReadHeader(ReadOnlySequence<byte> buffer, out bool compressed, out int messageLength)
+        {
+            if (buffer.Length < HeaderSize)
+            {
+                compressed = false;
+                messageLength = 0;
+                return false;
+            }
+
+            if (buffer.First.Length >= HeaderSize)
+            {
+                var headerData = buffer.First.Span.Slice(0, HeaderSize);
+
+                compressed = ReadCompressedFlag(headerData[0]);
+                messageLength = DecodeMessageLength(headerData.Slice(1));
+            }
+            else
+            {
+                Span<byte> headerData = stackalloc byte[HeaderSize];
+                buffer.Slice(0, HeaderSize).CopyTo(headerData);
+
+                compressed = ReadCompressedFlag(headerData[0]);
+                messageLength = DecodeMessageLength(headerData.Slice(1));
+            }
+
+            return true;
+        }
+
+        private static bool ReadCompressedFlag(byte flag)
+        {
+            if (flag == 0)
+            {
+                return false;
+            }
+            else if (flag == 1)
+            {
+                return true;
+            }
+            else
+            {
+                throw new InvalidOperationException("Unexpected compressed flag value in message header.");
+            }
+        }
+
+        /// <summary>
+        /// Read messages from the pipe reader.
+        /// </summary>
+        /// <param name="input">The request pipe reader.</param>
+        /// <param name="supportMultipleMessages">
+        /// A value that indicates whether multiple messages can be read from the reader.
+        /// When false, a complete message must be read from the reader and the reader is completed with no further data.
+        /// When true, a complete message must be read from the reader, and the reader can be open with additional data.
+        /// </param>
+        /// <returns>Complete message data.</returns>
+        public static async ValueTask<byte[]> ReadMessageAsync(this PipeReader input, bool supportMultipleMessages)
```

Comment:
I think we should have 2 methods. One that reads a single message and one that reads mulitple (basically `IAsyncEnumerable<byte[]>`)

Reply:
Do you want the read multiple to use C# 8 async + yield and return an `IAsyncEnumerable<byte[]>`?

I think it is something to try out. Can it wait though? I'd like to avoid requiring VS2019 just yet. Plus it would be good to compare perf with the current implementation.

Reply:
I hate how complicated this method is 😢

Reply:
I have split it: `ReadMessageAsync` and `ReadMessageStreamAsync`